### PR TITLE
22942 update model to use new sql_versioning library

### PR DESCRIPTION
**Pull request by kzdev420**
*Issue #:* /bcgov/entity#22942

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: legal-api/src/legal_api/models/amalgamation.py [hunk 1]
```diff
@@ -97,10 +97,10 @@ def get_revision_by_id(cls, transaction_id, amalgamation_id):
     def get_revision(cls, transaction_id, business_id):
         """Get amalgamation for the given transaction id."""
         # pylint: disable=singleton-comparison;
-        amalgamation_version = version_class(Amalgamation)
+        amalgamation_version = VersioningProxy.version_class(db.session(), Amalgamation)
         amalgamation = db.session.query(amalgamation_version) \
             .filter(amalgamation_version.transaction_id <= transaction_id) \
-            .filter(amalgamation_version.operation_type == 0) \
+            .filter(amalgamation_version.operation_type == 1) \
```
Comment (kzdev420) [inline legal-api/src/legal_api/models/amalgamation.py:103]: Common Values for operation_type:
INSERT: 1, UPDATE: 2, DELETE: 3

But, looks like the old_version used 0, 1, 2.

Comment (argush3) [reply legal-api/src/legal_api/models/amalgamation.py:102]: We should still be using the old db versioning approach of `INSERT: 0, UPDATE: 1, DELETE: 2`.  If this is not the case then the sql_versioning library will need to be updated.

Comment (argush3) [reply legal-api/src/legal_api/models/amalgamation.py:102]: I took a look at [sql_versioning's versioning.py](https://github.com/bcgov/lear/blob/4b41a96e4af2420d98694bc79d24641b85251983/python/common/sql-versioning/sql_versioning/versioning.py#L119) and the [insert/update/delete tests](https://github.com/bcgov/lear/blob/4b41a96e4af2420d98694bc79d24641b85251983/python/common/sql-versioning/tests/test_versioning.py) for sql_versioning library.  It looks like the old versioning operation types of `insert: 0, update: 1, delete: 2` are used in the new db versioning.

Are you sure you are seeing `insert: 1, update: 2, delete: 3` being used in the new versioning?

Comment (kzdev420) [reply legal-api/src/legal_api/models/amalgamation.py:102]: I just had the operations(Insert, Update, Delete).
The new_version used `insert: 1, update: 1, delete: 2`

Comment (argush3) [reply legal-api/src/legal_api/models/amalgamation.py:102]: please provide context of where you found this data?  did you create it yourself?  and if you created it yourself, what did you do to get this data?

Comment (kzdev420) [reply legal-api/src/legal_api/models/amalgamation.py:102]: I created the sample endpoint myself for confirmation.


```
@bp.route('/parties', methods=['POST'])
@bp.route('/parties/<int:party_id>', methods=['POST', 'PUT'])
@cross_origin(origin='*')
@jwt.requires_auth
def post_parties(party_id=None):
    print('hello world')
    json_input = copy.deepcopy(request.get_json())
    if party_id:
        party = Party.find_by_id(party_id)
        party.last_name = 'updated_ last name'
    else:
        party = Party(
            first_name='first name',
            last_name='last name',
            delivery_address_id=23718,
            mailing_address_id=23719
        )
    party.save()

    return jsonify(json_input), \
        (HTTPStatus.CREATED if (request.method == 'POST') else HTTPStatus.ACCEPTED)


@bp.route('/parties/<int:party_id>', methods=['DELETE'])
@cross_origin(origin='*')
@jwt.requires_auth
def delete_parties(party_id=None):
    print('hello world')
    json_input = copy.deepcopy(request.get_json())
    if party_id:
        party = Party.find_by_id(party_id)
        party.last_name = 'updated_ last name'
    
        party.delete()
        return jsonify({'message': 'Party has been deleted.'}), HTTPStatus.OK
```

Comment (kzdev420) [reply legal-api/src/legal_api/models/amalgamation.py]: Fixed.
Should comment the make_versioned function out.

File: legal-api/flags.json [hunk 2]
```diff
@@ -11,17 +11,22 @@
...
+                "filer": false,
+                "entity-bn": false,
+                "digital-credentials": false,
+                "dissolutions-job": false,
+                "furnishings-job": false,
+                "emailer-reminder-job": false,
+                "future-effective-job": false,
+                "update-colin-filings-job": false,
+                "update-legal-filings-job": false
+            }
```
Comment (chenhongjing) [inline legal-api/flags.json:29]: The old json-based FF format is updated not following the original design. You don't need to change it.

Comment (kzdev420) [reply legal-api/flags.json]: Reverted

**Comment by argush3** [comment]
Please provide some examples of what kind of versioning data you are seeing for create, update and delete scenarios.  I want to see that versioning is working when new db versioning is off and when it's on with these changes you've made.

**Comment by chenhongjing** [comment]
Are you going to do bcgov/entity/issues/22943 in a separate PR? I think updates are similar and you can update them together.

**Comment by chenhongjing** [comment]
> Please provide some examples of what kind of versioning data you are seeing for create, update and delete scenarios. I want to see that versioning is working when new db versioning is off and when it's on with these changes you've made.
> 
> Also, whenever possible, provide a summary of what you've updated in the PR description.

Following Argus' comment
Can you also update [this flag](https://github.com/bcgov/lear/blob/main/legal-api/flags.json#L14) both true & false and check? I'm wondering if the new versioning and versioning switching between old and new will break some tests.

**Comment by chenhongjing** [comment]
> I am seeing these errors when set `FF.legal-api: true` up There are 2 main issues with Resolution and IA_Filing.
> 
> * Resolution issue
>   The Resolution model has a property named [resolution_type](https://github.com/bcgov/lear/blob/7e24e5a83db063f804195ac199fb375e33597061/legal-api/src/legal_api/models/resolution.py#L38) and it is defined the `type` at the original and versioning tables.
>   But, the versioning doesn't work with `resolution_type` property.
>   When I changed the `resolution_type` to `type` in the model, it worked.
> * `test_business_filing_ia_parties`: I didn't found the issue points. :(

1. For the first issue, you can do any changes to fix the bug instead of just pinging me so that everyone can check if the changes are feasible
2. The second issue may be from `VersionedBusinessDetailsServices` you modified from the error I got when running your code

ps. If you want to show the failed unit test in CI, turn the db-versioning/legal-api on

File: queue_services/entity-digital-credentials/tests/unit/__init__.py
Comment (leodube-aot) [inline queue_services/entity-digital-credentials/tests/unit/__init__.py:1]: Updating digital credentials will be done in a separate ticket, so I think you can undo the changes in this file for now. Should resolve that failing digital credentials unit test as well.

**Comment by chenhongjing** [comment]
Like I said before, you can change FF `db-versioning/legal-api: true` to show us the CI result. Now we can only see the result when flag is off.

**Comment by kzdev420** [comment]
**Used the updated Versioning library.**

- Get the version class
  old: `sqlalchemy_continuum.version_class(model)`
  new: `VersioningProxy.version_class(db.session(), model)`

- Create a new transaction
  old:
  ```
   uow = versioning_manager.unit_of_work(session)
   transaction = uow.create_transaction(session)
  ```
  new: `transaction_id = VersioningProxy.get_transaction_id(session())` 


**Example of new versioning library**
 ```
Created new transaction: 4
Set current_transaction_id: 4

Creating version for Business (id=4), operation_type: I
Version created/updated for Business (id=4), transaction_id: 4

Poping out existing transaction: 4
Created new transaction: 5
Set current_transaction_id: 5

_Create a new Filing, filing.transaction_id: 5_
 
Creating version for Party (id=1), operation_type: I
Version created/updated for Party (id=1), transaction_id: 5

Creating version for Address (id=1), operation_type: I
Version created/updated for Address (id=1), transaction_id: 5

Creating version for PartyRole (id=1), operation_type: I
Version created/updated for PartyRole (id=1), transaction_id: 5

Creating version for Party (id=2), operation_type: I
Version created/updated for Party (id=2), transaction_id: 5

Creating version for Business (id=4), operation_type: U
Version created/updated for Business (id=4), transaction_id: 5

Exiting VersioningProxy.unlock_versioning()
Exiting setup_versioning.<locals>.clear_transaction()
```

**Comment by argush3** [comment]
@kzdev420 as per sprint planning yesterday, please close this PR

**Comment by kzdev420** [comment]
This PR is closed because the flask upgrade is on hold for now.

**Pull request closed without merging**